File: ACM/Acwing/135/135.py
'''
Author: NEFU AB-IN
Date: 2022-02-27 16:35:09
FilePath: \ACM\Acwing\135.py
LastEditTime: 2022-02-27 16:46:28
'''

from collections import deque

N = int(3e5 + 100)
INF = int(2e9)
a = [0] * N

if __name__ == "__main__":
    n, m = map(int, input().split())
    a[1:] = list(map(int, input().split()))
    q = deque()
    res = -INF
    for i in range(1, n + 1):
        a[i] += a[i - 1]
    q.append(0)  # 左端点初始值为0
    for i in range(1, n + 1):  #枚举右端点
        while q and i - q[0] > m:
            q.popleft()
        res = max(res, a[i] - a[q[0]])
        while q and a[q[-1]] >= a[i]:
            q.pop()
        q.append(i)
    print(res)

File: ACM/Acwing/135/135.cpp
/*
 * @Author: NEFU AB-IN
 * @Date: 2023-02-23 21:36:31
 * @FilePath: \Acwing\135\135.cpp
 * @LastEditTime: 2023-02-23 21:58:40
 */
#include <bits/stdc++.h>
using namespace std;
#define int long long
#undef int

#define SZ(X) ((int)(X).size())
#define ALL(X) (X).begin(), (X).end()
#define IOS                                                                                                            \
    ios::sync_with_stdio(false);                                                                                       \
    cin.tie(nullptr);                                                                                                  \
    cout.tie(nullptr)
#define DEBUG(X) cout << #X << ": " << X << '\n'
typedef pair<int, int> PII;

const int N = 3e5 + 10, INF = 0x3f3f3f3f;
int a[N], s[N];

signed main()
{
    IOS;
    int n, m;
    cin >> n >> m;

    deque<int> q;
    for (int i = 1; i <= n; ++i)
    {
        cin >> a[i];
        s[i] = s[i - 1] + a[i];
    }

    q.push_back(0); // 队列一定要有值，不然q.front无法取
    int res = -INF;
    for (int i = 1; i <= n; ++i)
    {
        while (SZ(q) && i - q.front() > m)
            q.pop_front();
        res = max(res, s[i] - s[q.front()]);
        while (SZ(q) && s[q.back()] >= s[i])
            q.pop_back();
        q.push_back(i);
    }

    cout << res;
    return 0;
}